### PR TITLE
Add modular increment and decrement arithmetic primitives

### PR DESCRIPTION
**Pull request by ksk-jij**
## Summary
- Add X-gate modular increment/decrement qkernels under `qamomile.circuit.algorithm.arithmetic` and re-export them from `qamomile.circuit.arithmetic`.
- Keep controlled variants as user-composed operations via `qmc.control(modular_increment)` and `qmc.control(modular_decrement)` rather than adding dedicated controlled helper APIs.
- Normalize nested controlled/composite blocks with slice markers through borrow checking before stripping emit-only markers, including the base Qiskit fallback path and backend overrides.
- Expand arithmetic coverage for sampler and expval paths, explicit wrap-around/control boundary cases, and explicit backend skips/xfails for known SDK limitations.

## Tests
- `uv run ruff format qamomile/ tests/`
- `uv run ruff check qamomile/ tests/`
- `uv run zuban check qamomile/`
- `uv run pytest tests/circuit/algorithm/arithmetic/test_modular.py -q -m ""`
- `uv run pytest tests/circuit/test_controlled.py::TestControlledVectorSubArgFollowUpOps tests/circuit/algorithm/arithmetic/test_modular.py -q -m ""`
- `uv run pytest -q`

### CODE REVIEW

File: qamomile/circuit/algorithm/arithmetic/modular_incdec.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    n = q.shape[0]
+    for k in qmc.range(n):
+        target_index = n - 1 - k
+        mcx = qmc.control(qmc.x, num_controls=target_index + 1)
+        prefix = q[0:target_index]
+        target = q[target_index]
+        control, prefix, target = mcx(control, prefix, target)
+        q[0:target_index] = prefix
+        q[target_index] = target
+    return control, q
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply qamomile/circuit/algorithm/arithmetic/modular_incdec.py]: Fixed in `3ddfb089`: the controlled variants now special-case `target_index == 0` with a direct `CX`, so no empty prefix slice is passed to the controlled-X wrapper.

File: qamomile/circuit/algorithm/arithmetic/modular_incdec.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    """
+    n = q.shape[0]
+    for target_index in qmc.range(n):
+        mcx = qmc.control(qmc.x, num_controls=target_index + 1)
+        prefix = q[0:target_index]
+        target = q[target_index]
+        control, prefix, target = mcx(control, prefix, target)
+        q[0:target_index] = prefix
+        q[target_index] = target
+    return control, q
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply qamomile/circuit/algorithm/arithmetic/modular_incdec.py]: Fixed in `3ddfb089`: the controlled variants now special-case `target_index == 0` with a direct `CX`, so no empty prefix slice is passed to the controlled-X wrapper.

File: qamomile/circuit/algorithm/arithmetic/modular_incdec.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        target_index = num_system - 1 - k
+        mcx = qmc.control(qmc.x, num_controls=target_index + 1)
+        control = q[control_index]
+        prefix = q[0:target_index]
+        target = q[target_index]
+        control, prefix, target = mcx(control, prefix, target)
+        q[control_index] = control
+        q[0:target_index] = prefix
+        q[target_index] = target
+    return q
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply qamomile/circuit/algorithm/arithmetic/modular_incdec.py]: Fixed in `3ddfb089`: the controlled variants now special-case `target_index == 0` with a direct `CX`, so no empty prefix slice is passed to the controlled-X wrapper.

File: qamomile/circuit/algorithm/arithmetic/modular_incdec.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    for target_index in qmc.range(num_system):
+        mcx = qmc.control(qmc.x, num_controls=target_index + 1)
+        control = q[control_index]
+        prefix = q[0:target_index]
+        target = q[target_index]
+        control, prefix, target = mcx(control, prefix, target)
+        q[control_index] = control
+        q[0:target_index] = prefix
+        q[target_index] = target
+    return q
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply qamomile/circuit/algorithm/arithmetic/modular_incdec.py]: Fixed in `3ddfb089`: the controlled variants now special-case `target_index == 0` with a direct `CX`, so no empty prefix slice is passed to the controlled-X wrapper.

File: qamomile/circuit/algorithm/arithmetic/modular_incdec.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
+"""Modular increment and decrement operators for basis-state registers."""
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply qamomile/circuit/algorithm/arithmetic/modular_incdec.py:1]: Fixed in f17a3cc0: the PR description now states that the exported API is the plain modular increment/decrement primitives and that controlled usage is composed with `qmc.control(...)`, not separate controlled/indexed helpers.

File: qamomile/circuit/transpiler/passes/emit_support/controlled_emission.py [hunk 6]
```diff
@@ -1118,6 +1121,35 @@ def blockvalue_to_gate(
...
+        ``Block``; otherwise the original object unchanged.
+    """
+    if not isinstance(block_value, Block):
+        return block_value
+
+    from qamomile.circuit.transpiler.passes.strip_slice_ops import (
+        StripSliceArrayOpsPass,
+    )
+
+    return StripSliceArrayOpsPass().run(block_value)
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply qamomile/circuit/transpiler/passes/emit_support/controlled_emission.py]: Fixed in f17a3cc0: nested controlled/composite blocks that contain slice markers now run constant folding with markers preserved, then `SliceBorrowCheckPass`, then marker stripping before gate conversion or fallback emit.

**Comment by ksk-jij** [comment]
## English Branch Summary

This summary reflects the final merged state of PR #453, including the removal of the temporary `qamomile.circuit.arithmetic` shim and the final public import path through `qamomile.circuit`.

### 1. Problem Overview

This branch adds modular increment and decrement primitives for computational-basis qubit registers to Qamomile's circuit algorithm layer. `modular_increment` maps a little-endian `Vector[Qubit]` as `|j> -> |j + 1 mod 2^n>`, and `modular_decrement` maps it as `|j> -> |j - 1 mod 2^n>`. The implementation is deliberately X-gate based: it uses a ripple of X gates and multi-controlled X gates rather than a QFT-based adder, so the primitive does not depend on whether a backend can lower QFT composites.

Before this branch, Qamomile did not provide a small public modular arithmetic building block that users could import directly from the circuit API. This made patterns such as Quantum Eigenvalue Transformation, quantum walks, or any circuit that conditionally advances a system register by one require users to manually write the MCX ripple each time. The branch provides the primitive as ordinary increment/decrement kernels, and controlled variants are built with the existing `qmc.control(...)` composition mechanism.

The branch also fixes controlled emission paths for nested sub-kernels that contain slices. Top-level qkernel blocks already go through slice borrow checking and slice marker stripping before emission, but nested `Block` objects stored inside `ControlledUOperation.block` or custom composite `implementation_block` fields are not directly visited by the normal top-level pass flow. On main, those nested blocks could reach gate conversion or fallback emission with slice markers still present, which could either bypass borrow checking or surface as backend-specific `SliceArrayOperation` runtime errors. This branch normalizes nested blocks that contain slice markers by running constant folding, `SliceBorrowCheckPass`, and `StripSliceArrayOpsPass` before gate conversion or fallback emission.

### 2. Frontend and Public API Changes

Users can import `modular_increment` and `modular_decrement` directly from `qamomile.circuit`. The final public path is `from qamomile.circuit import modular_increment, modular_decrement`. The implementation lives under `qamomile.circuit.algorithm.arithmetic`, and `qamomile.circuit` re-exports the primitives from the implementation module so the generated API reference can show them on the main circuit API page. The temporary `qamomile.circuit.arithmetic` shim was removed.

```python
import qamomile.circuit as qmc
from qamomile.circuit import modular_decrement, modular_increment

@qmc.qkernel
def step(q: qmc.Vector[qmc.Qubit], direction: qmc.UInt) -> qmc.Vector[qmc.Qubit]:
    if direction == qmc.uint(0):
        q = modular_increment(q)
    else:
        q = modular_decrement(q)
    return q
```

No dedicated `controlled_modular_increment` or `controlled_modular_decrement` API was added. Users should pass the ordinary primitive to `qmc.control(...)`, which keeps the API surface small and lets the existing controlled-gate machinery handle control qubits and multi-control variants.

```python
import qamomile.circuit as qmc
from qamomile.circuit import modular_increment

@qmc.qkernel
def controlled_step(control: qmc.Qubit, q: qmc.Vector[qmc.Qubit]) -> tuple[qmc.Qubit, qmc.Vector[qmc.Qubit]]:
    controlled_increment = qmc.control(modular_increment, num_controls=1)
    control, q = controlled_increment(control, q)
    return control, q
```

The qubit vector is interpreted as little-endian: `q[0]` is the least-significant bit. For concrete zero-length registers, the public wrapper raises a `ValueError` containing `zero-length register` instead of letting a raw `IndexError` escape from `q[0]` access.

The controlled-gates tutorial was also updated around sliced controlled sub-kernels. The documentation now treats sliced controlled QFT-style cases as supported where block-to-gate conversion succeeds, while making clear that backends without block-to-gate conversion may still reject the multi-target fallback path.

### 3. Backend and IR Changes

`modular_increment` and `modular_decrement` are implemented in `qamomile/circuit/algorithm/arithmetic/modular_incdec.py`. Increment iterates over the higher bits with `qmc.range(1, n)`, applies prefix-controlled X gates to propagate the carry, and flips the LSB at the end. Decrement flips the LSB first and then walks target indices upward with the same prefix-controlled X pattern. Both are built from `qmc.x` and `qmc.control(qmc.x, num_controls=...)` only.

Controlled emission now has shared nested-block preparation helpers in `controlled_emission.py`. `_contains_slice_markers` recursively detects `SliceArrayOperation` and `ReleaseSliceViewOperation` not only in top-level operations, but also inside control-flow children, nested `block` fields, and nested `implementation_block` fields. `_prepare_nested_block_for_emit` only acts when markers are present; it treats traced nested blocks as hierarchical for pass execution, then runs constant folding, slice borrow checking, and slice marker stripping.

The preparation helper is wired into the major controlled/composite entry points: symbolic controlled-U, multi-argument controlled-U, concrete controlled-U, QPE controlled-power fallback, base controlled fallback, custom composite emission, and block-to-gate conversion. Additional recursive helpers prepare nested operation block fields so nested `ControlledUOperation` and custom composite blocks receive the same normalization even when they appear inside another nested block.

The base controlled fallback now rejects unsupported multi-control or general multi-target inner blocks explicitly with `EmitError` rather than attempting a gate-by-gate fallback that could silently miscompile. The only multi-target exception kept in the fallback is a single SWAP, because that path explicitly maps both target indices.

Tests were added for plain and controlled modular increment/decrement across several register sizes and seeds, including sampling and expectation-value paths. Boundary cases explicitly cover wrap-around and control enabled/disabled behavior, rather than relying on random inputs to hit those cases. The backend fixture logic was consolidated into `tests/circuit/conftest.py` so SDK availability and backend labeling are shared instead of being copied inside the modular arithmetic tests.

Regression tests were also added for nested slice marker handling. These cover nested controlled slice borrow violations, base controlled fallback after forced block-to-gate failure, QPE controlled-power fallback with sliced nested blocks, and sliced controlled sub-kernel sampling across SDK backends where supported.

### 4. API Reference State

`docs/api_gen/toc.py` is intentionally back to the same implementation as `main`; no special TOC exclusion is needed anymore. Because the `qamomile.circuit.arithmetic` package was removed, the API generator no longer discovers or emits `api/circuit/arithmetic`. A regenerated API reference shows `modular_increment` and `modular_decrement` on the `qamomile.circuit` page and still shows the algorithm-level exports under `qamomile.circuit.algorithm`.

### 5. Alternatives Not Chosen

A QFT-based increment/decrement implementation was not chosen. Although QFT adders are natural for arithmetic circuits, the goal here was to provide a small X-based building block that does not introduce a new dependency on QFT composite lowering or backend-specific QFT support.

Dedicated controlled APIs such as `controlled_modular_increment` and `controlled_modular_decrement` were not added. Qamomile already has `qmc.control(...)`, and using the ordinary primitive with that API avoids adding another public surface while keeping the user-facing controlled form consistent with the rest of the circuit API.

The repeated MCX ripple pattern was not factored into a private qkernel helper. That would reduce visual duplication, but it would also change the nested-call IR shape and the controlled-emission behavior in a part of the implementation where stability matters more than small local deduplication. The current implementation keeps the increment and decrement kernels simple and locally readable.

Nested slice marker handling was placed in shared controlled-emission helpers rather than backend-specific overrides. Putting the fix in individual backend paths would make future backends or future controlled/composite entry points easy to miss. At the same time, the helper only runs the expensive normalization when slice markers are actually detected, so slice-free controlled modular arithmetic is not unnecessarily transformed.

Zero-length registers raise `ValueError` rather than becoming no-ops. A zero-qubit modular shift can be viewed mathematically as identity, but the current qkernel tracing model still evaluates the syntax that indexes `q[0]`. Rejecting concrete zero-length registers at the public wrapper gives users a clear error and avoids carrying a special no-op shape into IR lowering.

### 6. Known Limitations

QURI Parts support for modular arithmetic is partial. Plain modular arithmetic is skipped for registers larger than two qubits because the current QURI Parts emitter cannot emit the multi-controlled X gates produced by the ripple construction. This should become a normal assertion once QURI Parts gains native MCX support or a decomposition that preserves the required structure.

QURI Parts support for controlled modular arithmetic is also partial. Controlled modular shifts over multi-qubit registers still hit loop-local bindings and classical `BinOp` shapes that the current recursive controlled emitter does not fully handle. This remains an explicit test skip rather than an accidental green path.

CUDA-Q sampling supports the controlled modular shape used in the tests, but the CUDA-Q observe path currently rejects the runtime-control-flow shape used for controlled modular arithmetic. The expectation-value path is therefore marked with an explicit `pytest.xfail` for that known CUDA-Q limitation.

The base controlled fallback intentionally rejects general multi-control or multi-target inner blocks when block-to-gate conversion fails. This prevents silent miscompilation. A future lower-level fallback emitter with a full block-local operand-to-target mapping could expand the supported fallback cases.

Future QPE-style or manual composite emitters can reintroduce slice-marker normalization bugs if they bypass the shared controlled/composite entry points and directly emit block-valued operands. A longer-term hardening path would be to move nested-block preparation closer to the lowest block-valued emission helper.

`_prepare_nested_block_for_emit` is safest when callers provide bindings that resolve vector shapes and slice boundaries. Dominant controlled-U paths already bind quantum input shapes before calling it, but direct helper use with unresolved symbolic intervals can still leave correctness dependent on caller context. A future improvement would be stronger symbolic interval handling inside `SliceBorrowCheckPass` itself.
